source/stream.c

```diff
@@ -0,0 +1,359 @@
+             */
+            if (checked_offset > 0 || checked_offset == INT64_MIN ||
+                (uint64_t)(-checked_offset) > (uint64_t)impl->original_cursor.len) {
+                return aws_raise_error(AWS_IO_STREAM_INVALID_SEEK_POSITION);
+            }
+
+            /* cases that would make this unsafe became errors with previous conditional */
+            final_offset = (uint64_t)impl->original_cursor.len - (uint64_t)(-checked_offset);
+            break;
+    }
+
+    /* true because we already validated against (impl->original_cursor.len) which is <= SIZE_MAX */
+    assert(final_offset <= SIZE_MAX);
+
+    impl->current_cursor = impl->original_cursor;
+
+    /* (size_t) final_offset - safe via previous assert's reasoning */
+    aws_byte_cursor_advance(&impl->current_cursor, (size_t)final_offset);
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_aws_input_stream_byte_cursor_read(
+    struct aws_input_stream *stream,
+    struct aws_byte_buf *dest,
+    size_t *amount_read) {
+    struct aws_input_stream_byte_cursor_impl *impl = stream->impl;
+
+    *amount_read = 0;
+
+    size_t actually_read = dest->capacity - dest->len;
+    if (actually_read > impl->current_cursor.len) {
+        actually_read = impl->current_cursor.len;
+    }
+
+    if (!aws_byte_buf_write(dest, impl->current_cursor.ptr, actually_read)) {
+        return AWS_OP_ERR;
+    }
+
+    aws_byte_cursor_advance(&impl->current_cursor, actually_read);
+
+    *amount_read = actually_read;
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_aws_input_stream_byte_cursor_get_status(
+    struct aws_input_stream *stream,
+    struct aws_stream_status *status) {
+    struct aws_input_stream_byte_cursor_impl *impl = stream->impl;
+
+    status->is_end_of_stream = impl->current_cursor.len == 0;
+    status->is_valid = true;
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_aws_input_stream_byte_cursor_get_length(struct aws_input_stream *stream, size_t *out_length) {
+    struct aws_input_stream_byte_cursor_impl *impl = stream->impl;
+
+    *out_length = impl->original_cursor.len;
+
+    return AWS_OP_SUCCESS;
+}
+
+static void s_aws_input_stream_byte_cursor_clean_up(struct aws_input_stream *stream) {
+    struct aws_input_stream_byte_cursor_impl *impl = stream->impl;
+
+    aws_mem_release(stream->allocator, impl);
+}
+
+static struct aws_input_stream_vtable s_aws_input_stream_byte_cursor_vtable = {
+    .seek = s_aws_input_stream_byte_cursor_seek,
+    .read = s_aws_input_stream_byte_cursor_read,
+    .get_status = s_aws_input_stream_byte_cursor_get_status,
+    .get_length = s_aws_input_stream_byte_cursor_get_length,
+    .clean_up = s_aws_input_stream_byte_cursor_clean_up};
+
+struct aws_input_stream *aws_input_stream_new_from_cursor(
+    struct aws_allocator *allocator,
+    const struct aws_byte_cursor *cursor) {
+    struct aws_input_stream *input_stream = aws_mem_acquire(allocator, sizeof(struct aws_input_stream));
```

Comment:
we could do a single aws_mem_acquire with enough memory for both structs. `aws_mem_acquire_many()` exists for that purpose.

Or achieve the same thing by having the aws_input_stream_byte_cursor_impl contain the aws_input_stream

Reply:
We allocate polymorphic things in a single block elsewhere?  Haven't noticed that.

Reply:
Colden has used aws_mem_acquire_many() a few times in aws-c-mqtt

I've been making "subclasses" that contain the base type as a member. I personally feel like that's simpler to implement.